lib/wallet/test/unit/Cardano/Wallet/CheckpointsSpec.hs

```diff
@@ -1,268 +1,120 @@
+import Cardano.Wallet.Checkpoints.Policy
+    ( sparseArithmetic )
+import Cardano.Wallet.Gen
+    ( genSlotNo )
+import Cardano.Wallet.Primitive.Types
+    ( Slot, SlotNo (..), WithOrigin (..) )
+import Data.Delta
+    ( Delta (..) )
 import Test.Hspec
-    ( Spec, describe, it, shouldBe )
+    ( Spec, describe, it )
 import Test.QuickCheck
     ( Arbitrary (..)
-    , Gen
     , Property
     , choose
-    , conjoin
-    , counterexample
     , forAll
+    , frequency
+    , getPositive
+    , listOf
     , property
-    , (.&&.)
     , (===)
-    , (==>)
     )
 
-import qualified Data.List as L
-
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.Strict as Map
 
 spec :: Spec
 spec = do
-    describe "sparseCheckpoints" $ do
-        it "k=2160, h=42" $ \_ -> do
-            let cfg = SparseCheckpointsConfig
-                    { edgeSize = 10
-                    , epochStability = 2160
-                    }
-            let h = Quantity 42
-
-            -- First unstable block: 0
-            sparseCheckpoints cfg h `shouldBe`
-                [ 0
-                , 32,33,34,35,36,37,38,39,40,41 -- Short-term checkpoints
-                , 42 -- Tip
-                ]
-
-        it "k=2160, h=2414" $ \_ -> do
-            let cfg = SparseCheckpointsConfig
-                    { edgeSize = 10
-                    , epochStability = 2160
-                    }
-            let h = Quantity 2714
-            -- First unstable block: 554
-            sparseCheckpoints cfg h `shouldBe`
-                [ 0
-                , 720, 1440, 2160               -- Long-term checkpoints
-
-                , 2704, 2705, 2706, 2707, 2708  -- Short-term checkpoints
-                , 2709, 2710, 2711, 2712, 2713  -- edgeSize = 10
-
-                , 2714  -- Tip
-                ]
-
-        it "k=2160, h=2414" $ \_ -> do
-            let cfg = SparseCheckpointsConfig
-                    { edgeSize = 0
-                    , epochStability = 2160
-                    }
-            let h = Quantity 2714
-            -- First unstable block: 554
-            sparseCheckpoints cfg h `shouldBe`
-                [ 0
-                , 720, 1440, 2160               -- Long-term checkpoints
-                , 2714  -- Tip
-                ]
-
-        it "The tip is always a checkpoint" $ \_ ->
-            property prop_sparseCheckpointTipAlwaysThere
-
-        it "There's at least (min h edgeSize) checkpoints" $ \_ ->
-            property prop_sparseCheckpointMinimum
-
-        it "∀ cfg. sparseCheckpoints (cfg { edgeSize = 0 }) ⊆ sparseCheckpoints cfg" $ \_ ->
-            property prop_sparseCheckpointEdgeSize0
-
-        it "Checkpoints are eventually stored in a sparse manner" $ \_ ->
-            property prop_checkpointsEventuallyEqual
+    describe "extendAndPrune" $ do
+        it "actually prunes checkpoints" $
+            property prop_doesPrune
+        it "keeps the tip of the chain" $
+            property prop_keepTip
```

Comment:
if I understand well we are always going to keep 
(a) genesis
(b) latest tip
(c) some checkpoints ONLY within stability epoch (I mean this range defined by security parameter that limits depth of rollbacks)
What about adding properties checking that : 
(a) genesis cp is always there 
(b) after pruning we can only have at most genesis (either it is in stability window or not) and checkpoints within the stability window 
(c) variate stability window explicitly ((i) all possible cps are within it, (ii) stability window is proper subset of all possible checkpoints)

Reply:
>  What about adding properties checking that :

Well, these properties are properties of the specific `policy` argument, not of the `extendAndPrune` function. It's enough to test that `extendPrune` does *use* the policy, and then rely on the properties of the specific policy to get a specific result. The specific properties are tested here:

https://github.com/input-output-hk/cardano-wallet/blob/0c2382c373f76e808fd3472b78d1ce98c7770618/lib/wallet/test/unit/Cardano/Wallet/Checkpoints/PolicySpec.hs#L52-L66

I can add a test that `sparseArithmetic` does keep genesis.